### PR TITLE
Let SourceSet expose type-safely that it is ExtensionAware

### PR DESCRIPTION
**Pull request by eskatos**
So that usage of extra properties and extensions on `SourceSet` instances is possible without casting from statically compiled languages.

### CODE REVIEW

File: subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java [hunk 1]
```diff
@@ -247,4 +257,26 @@ public SourceSet resources(Action<? super SourceDirectorySet> configureAction) {
...
+    public Convention getConvention() {
+        assertExtensibleDynamicObject();
+        return extensibleDynamicObject.getConvention();
+    }
+
+    private void assertExtensibleDynamicObject() {
+        if (extensibleDynamicObject == null) {
+            extensibleDynamicObject = new ExtensibleDynamicObject(this, SourceSet.class, instantiator);
+        }
+    }
```
Comment (JLLeitschuh) [inline subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java]: Why aren't you doing this in the constructor the way that the `ProjectImpl` does?

Also, is there any concern about race conditions from multiple threads instantiating multiple instances of the `ExtensibleDynamicObject` for `this`?

Comment (eskatos) [reply subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java]: Right, let's make it simple first

File: subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java [hunk 2]
```diff
@@ -21,17 +21,23 @@
...
 import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.SourceSetOutput;
+import org.gradle.internal.metaobject.DynamicObject;
+import org.gradle.internal.reflect.Instantiator;
 import org.gradle.util.GUtil;
 
 import static org.gradle.util.ConfigureUtil.configure;
 
-public class DefaultSourceSet implements SourceSet {
+public class DefaultSourceSet implements SourceSet, DynamicObjectAware {
```
Comment (bamboo) [inline subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java]: ❓ Why `DynamicObjectAware`?

Comment (JLLeitschuh) [reply subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java]: If you implement `ExtensionAware` you must also implement `DynamicObjectAware`.
I don't know the historical reasons as to why, but I know that the injector won't construct your object if you don't have this type.

@oehme You seem to understand much of the historical significance of many decisions, any insights on this one?

https://github.com/gradle/gradle/blob/e975872ae2cf1d521b73f1107cae772ca09c63cd/subprojects/core/src/main/java/org/gradle/api/internal/AbstractClassGenerator.java#L121-L126

**Comment by oehme** [comment]
I'm not a huge fan of doing this for one specific type. It would be better if we worked towards clearer rules, e.g.

- if a type says it implements ExtensionAware, we automatically mix in the necessary machinery. All the user should have to do is make their class abstract and just don't implement those methods.
- if a type doesn't say it is ExtensionAware, we shouldn't be adding support for extensions implicitly

Of course this would require some deprecations.

The situation right now is confusing - Everything created by Gradle is in fact extension aware and you can cast it to that interface. Making some types statically define this makes it look like only they are extension aware though. And requiring internal APIs for statically typed extension awareness discourages plugin authors from making use of it.

**Comment by eskatos** [comment]
@oehme this was triggered by our user manual demonstrating the usage of _extra properties_ on source sets. And there's the fact that we want to move away from _source set conventions_ towards _source set extensions_.

Agreed that we need to go down the path of how plugin authors will leverage this on custom types (declaration and implementation wise). I think it should be statically declared in any case so that statically compiled languages get the feature exposed properly.

**Comment by JLLeitschuh** [comment]
While you're all here:
https://github.com/gradle/gradle/pull/6685

**Comment by JLLeitschuh** [comment]
> Everything created by Gradle is in fact extension aware and you can cast it to that interface.

This is not necessarily the case in the Kotlin DSL.
For example any of these `*Scope` object's will not expose the `ExtensionAware` interface (by casting) because they don't get decorated in by the Kotlin compiler.

https://github.com/gradle/kotlin-dsl/blob/76fb69a46878c2e523dbf09d5c0206a915832646/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt#L196-L264

That's why, for example, in this issue: #5546 in a `settings.gradle.kts` file:

**Fails at Runtime**
```kotlin
val kotlinVersion = (this as ExtensionAware).extra["kotlin.version"]
println(kotlinVersion)
```

**Compiles and Runs fine**

```kotlin
val kotlinVersion = (this.settings as ExtensionAware).extra["kotlin.version"]
println(kotlinVersion)
```

**Pull request merged**